plugins/arm_thumb/thumb.ml

```diff
@@ -0,0 +1,195 @@
+    | `tSTRHi, [dest; src; imm; unknown; _] -> 
+      lift_mem_single ~shift_val:1 dest src ~src2:imm St H
+    | `tSTRHr, [dest; src1; src2; unknown; _] -> 
+      lift_mem_single dest src1 ~src2 St H
+    | `tSTMIA, dest :: _dest :: _unknown :: _nil_reg :: src_list (* looks like they should be different, but actually are the same in Thumb mode *)
+    | `tSTMIA_UPD, dest :: _dest :: _unknown :: _nil_reg :: src_list ->
+      Mem.store_multiple dest src_list |> move
+    | `tLDMIA, dest :: _unknown :: _nil_reg :: src_list (* same as stmia *)
+    | `tLDMIA_UPD, dest :: _unknown :: _nil_reg :: src_list ->
+      Mem.load_multiple dest src_list |> move
+    | `tPUSH, _unknown :: _nil_reg :: src_list ->
+      Mem.push_multiple src_list |> move
+    | `tPOP, _unknown :: _nil_reg :: src_list ->
+      let has_pc = List.exists src_list 
+          (fun s -> match s with
+             |`Reg `PC -> true
+             | _ -> false) in
+      let pop_eff = Mem.pop_multiple src_list in
+      if has_pc then
+        ctrl (jmp (var Env.pc)) pop_eff addr
+      else move pop_eff
+    | _ -> move pass
+
+  let lift_bits insn ops =
+    let open Bits in
+    match insn, ops with
+    | `tSXTB, [|dest; src; _unknown; _|] -> sxtb dest src
+    | `tSXTH, [|dest; src; _unknown; _|] -> sxth dest src
+    | `tUXTB, [|dest; src; _unknown; _|] -> uxtb dest src
+    | `tUXTH, [|dest; src; _unknown; _|] -> uxth dest src
+    | _ -> []
+
+  (* these are not entirely complete *)
+  let lift_branch insn ops addr =
+    let open Defs in
+    let open Branch in
+    let addr = Core.int Env.value addr in
+    match insn, ops with
+    | `tBcc, [|target; cond; _cpsr|] -> tbcc cond target addr
+    | `tB, [|target; _unknown; _|] -> tb target addr
+    | `tBL, [|_unknown; _nil_reg; target; _cpsr|] -> tbl target
+    | `tBLXi, [|_unknown; _nil_reg; target; _cpsr|] -> tblxi target
+    | `tBLXr, [|_unknown; _nil_reg; target|] -> tblxr target
+    | `tBX, [|target; _unknown; _|] -> tbx target
+    | _ -> (skip, pass)
+
+  let lift_with (addr : Bitvec.t) (insn : Thumb_defs.insn)
+      (ops : Thumb_defs.op array) =
+    match insn with
+    | #move_insn -> lift_move insn ops |> DSL.expand |> move
+    | #mem_insn -> lift_mem insn ops addr
+    | #bits_insn -> lift_bits insn ops |> DSL.expand |> move
+    | #branch_insn -> 
+      let ctrl_eff, data_eff = lift_branch insn ops addr in
+      ctrl ctrl_eff data_eff addr (* var Env.pc *)
+
+end
+
+open Bap.Std
+
+let run_lifter _label addr insn _mem 
+    (lifter : Bitvec.t -> Defs.insn -> Defs.op array -> unit Theory.eff) =
+  match Insns.of_basic insn with
+  | None -> raise (Defs.Lift_Error "unknown instruction")
+  | Some arm_insn -> Disasm_expert.Basic.sexp_of_full_insn insn |> Sexp.to_string |> Stdio.print_endline;
+    match Insns.arm_ops (Disasm_expert.Basic.Insn.ops insn) with
+    | Error err -> raise (Defs.Lift_Error (Error.to_string_hum err))
+    | Ok ops -> lifter addr arm_insn ops
+
+let () =
```

Comment:
We missed an important check here, we need to request the `Arch.slot` and returns semantics only if the architecture is matching with thumb, otherwise this lifter will provide semantics for x86, mips, powerpc, arm, and any other architecture. The same is true for the arm lifter.

Reply:
Just added the check for both ARM and ARM Thumb